File: 1768. Merge Strings Alternately/solution.c
#include <stdio.h>
#include <assert.h>
#include <string.h>
#include <stdlib.h>
#include <stdbool.h>

char* mergeAlternately(char* word1, char* word2) {
    int read = 0;
    int write = 0;

    char* word = (char*)malloc(sizeof(char) * (strlen(word1) + strlen(word2) + 1));

    while (word1[read] != '\0' && word2[read] != '\0') {
        word[write++] = word1[read];
        word[write++] = word2[read];
        read++;
    }

    if (word1[read] == '\0') {
        while (word2[read] != '\0') {
            word[write++] = word2[read++];
        }
    } else {
        while (word1[read] != '\0') {
            word[write++] = word1[read++];
        }
    }
    word[write] = '\0';
    return word;
}


int main() {
    // First case.
    char word1[] = "abc";
    char word2[] = "pqr";

    // Second case.
    char word3[] = "ab";
    char word4[] = "pqrs";

    // Third case.
    char word5[] = "abcd";
    char word6[] = "pq";

    char* word = mergeAlternately(word1, word2);
    assert(strcmp(word, "apbqcr") == 0);
    
    word = mergeAlternately(word3, word4);
    assert(strcmp(word, "apbqrs") == 0);

    word = mergeAlternately(word5, word6);
    assert(strcmp(word, "apbqcd") == 0);
    free(word);
    return 0;
}

File: 1768. Merge Strings Alternately/solution.py
import re
from functools import reduce


def mergeAlternately(word1: str, word2: str) -> str:
    word1 = (
        word1 + ((len(word2) - len(word1)) * " ") if len(word1) < len(word2) else word1
    )
    word2 = (
        word2 + ((len(word1) - len(word2)) * " ") if len(word2) < len(word1) else word2
    )
    word = list(zip(word1, word2))

    return re.sub(" ", "", reduce(lambda acc, tpl: acc + "".join(tpl), word, ""))


if __name__ == "__main__":
    print(mergeAlternately("abc", "pqr"))
